src/containerbuddy/etcd.go

```diff
@@ -0,0 +1,240 @@
+		Prefix: "/containerbuddy",
+	}
+	etcdConfig := client.Config{}
+	switch endpoints := config["endpoints"].(type) {
+	case string:
+		etcdConfig.Endpoints = []string{endpoints}
+	case []string:
+		etcdConfig.Endpoints = endpoints
+	default:
+		log.Fatal("Must provide etcd endpoints")
+	}
+
+	prefix, ok := config["prefix"].(string)
+	if ok {
+		etcd.Prefix = prefix
+	}
+
+	etcdClient, err := client.New(etcdConfig)
+	if err != nil {
+		log.Fatal(err)
+	}
+	etcd.Client = etcdClient
+	etcd.API = client.NewKeysAPI(etcdClient)
+	return etcd
+}
+
+// Deregister removes this instance from the registry
+func (c Etcd) Deregister(service *ServiceConfig) {
+	c.deregisterService(service)
+}
+
+// MarkForMaintenance removes this instance from the registry
+func (c Etcd) MarkForMaintenance(service *ServiceConfig) {
+	c.deregisterService(service)
+}
+
+// SendHeartbeat refreshes the TTL of this associated etcd node
+func (c Etcd) SendHeartbeat(service *ServiceConfig) {
+	if c.checkServiceExists(service) {
+		c.updateServiceTTL(service)
+	} else {
+		log.Printf("Service not registered, registering...")
+		if err := c.registerService(service); err != nil {
+			log.Printf("Error registering service %s: %s", service.Name, err)
+		}
+
+	}
+}
+
+// CheckForUpstreamChanges checks another etcd node for changes
+func (c Etcd) CheckForUpstreamChanges(backend *BackendConfig) bool {
+	return c.checkHealth(backend)
+}
+
+func (c Etcd) getNodeKey(service *ServiceConfig) string {
+	return fmt.Sprintf("%s/%s/%s", c.Prefix, service.Name, service.ID)
+}
+
+func (c Etcd) getAppKey(appName string) string {
+	return fmt.Sprintf("%s/%s", c.Prefix, appName)
+}
+
+var etcdUpstreams = make(map[string][]EtcdServiceNode)
+
+func (c Etcd) checkHealth(backend *BackendConfig) bool {
+	services, err := c.getServices(backend.Name)
+	if err != nil {
+		if _, ok := err.(client.Error); !ok {
+			log.Printf("Failed to query %v: %s", backend.Name, err)
+		}
+		return false
+	}
+	didChange := etcdCompareForChange(etcdUpstreams[backend.Name], services)
+	if didChange || len(services) == 0 {
+		// We don't want to cause an onChange event the first time we read-in
+		// but we do want to make sure we've written the key for this map
+		etcdUpstreams[backend.Name] = services
+	}
+	return didChange
+}
+
+func (c Etcd) checkServiceExists(service *ServiceConfig) bool {
+	key := c.getNodeKey(service)
+	_, err := c.API.Get(context.Background(), key, nil)
+	if err != nil {
+		if etcdErr, ok := err.(client.Error); ok {
+			return etcdErr.Code != client.ErrorCodeKeyNotFound
+		}
+		log.Printf("Unexpected etcd Error on key %s: %s", key, err)
+	}
+	return true
+}
+
+func (c Etcd) getServices(appName string) ([]EtcdServiceNode, error) {
+	var services []EtcdServiceNode
+
+	key := c.getAppKey(appName)
+	resp, err := c.API.Get(context.Background(), key, &client.GetOptions{Recursive: true})
+	if err != nil {
+		if etcdErr, ok := err.(client.Error); ok {
+			if etcdErr.Code == client.ErrorCodeKeyNotFound {
```

Comment:
KeyNotFound errors are like 404's.  Need to check for them since they are not errors, just an indication that there are no service.

Use Case:

Backend health check is run before the backend has a chance to register itself with etcd. Rather than result in an error, it will return an empty array instead - which is more indicative of the actual state.